Added Date Format choice and Custom Min & Max Voltage

tmleafs opened the pull request:
Saw your suggestion of custom Min and Max voltage and while doing that I did date format as most DHs were different and I prefer the UK date format

![screenshot_20180122-124056 small](https://user-images.githubusercontent.com/8752233/35221321-9ab00f82-ff71-11e7-9f18-b42be3295868.jpg)

veeceeoh commented on devicetypes/bspranger/xiaomi-temperature-humidity-sensor.src/xiaomi-temperature-humidity-sensor.groovy:
```diff
@@ -134,7 +140,14 @@ metadata {
     // send event for heartbeat
-    def now = new Date().format("yyyy MMM dd EEE h:mm:ss a", location.timeZone)
+    def now
+	if(dateformat == "US" || dateformat == "" || dateformat == null)
```
I noticed that Brian had to change references of the variable `tempOffset` to `settings.tempOffset` in order to get conditionals / boolean operands to work correctly in the `parseTemperature()` function.

I haven't wrapped my head around global variables in Groovy, but doesn't that mean all references to `dateformat`, `voltsmax`, and `voltsmin` should be changed to `settings.dateformat`, `settings.voltsmax`, and `settings.voltsmin`?

veeceeoh commented:
Definitely keep the temperature offset setting in the least.

I can see potential for confusion if users can set a custom max / min voltage used for calculating battery percentage. If you make it clear that it's impossible to troubleshoot battery issues for users who have changed their settings, then I suppose it may be okay.

Having a UK date format setting seems fair enough. I don't understand why that isn't built into the hub's Groovy libraries like the timezone function is, though.

tmleafs commented:
> Having a UK date format setting seems fair enough. I don't understand why that isn't built into the hub's Groovy libraries like the timezone function is, though.

Thats due to freezing libraries and adding of any new features before the UK release of smartthings 

> Let's put the date code in a function and just call it instead of duplicating it a million times.

Yup, I agree it was me rushing as the last thing I started to fix/add someone beat me too the pull request

tmleafs commented:
`def formatDate(Boolean) {
	if(dateformat == "US" || dateformat == "" || dateformat == null){
    	if(Boolean == false || Boolean == "" || Boolean == null)
    	return (new Date().format("EEE MMM dd yyyy h:mm:ss a", location.timeZone))
    	else
   		return new Date().format("MMM dd yyyy", location.timeZone)
	}
	else if(dateformat == "UK"){
		if(Boolean == false || Boolean == "" || Boolean == null)
    	return new Date().format("EEE dd MMM yyyy h:mm:ss a", location.timeZone)
        else
        return new Date().format("dd MMM yyyy", location.timeZone)
    }
	else{
    	if(Boolean == false || Boolean == "" || Boolean == null)
        return new Date().format("EEE yyyy MMM dd h:mm:ss a", location.timeZone)
        else
        return new Date().format("yyyy MMM dd", location.timeZone)
	}
}`

Can any of you improve on this?

tmleafs commented:
Added format date function to all DH's
&
Added more to the settings page
![screenshot_20180123-131908 small](https://user-images.githubusercontent.com/8752233/35278176-14ce9d04-0041-11e8-9c60-cd8347f434c9.jpg)

bspranger commented:
I am still confused if this statement is correct:

 if(voltsmin != null || voltsmin == "")

Shouldn't it be:

 if(voltsmin != null && voltsmin != "")

tmleafs commented:
It's a OR just to rule out any reason it's not returning null the != "" shouldn't be needed I just put it in there to be safe

veeceeoh commented:
`if(voltsmin != null || voltsmin == "")`

 _if voltsmin is not equal to or voltsmin is equal to ""_ would yield _true_ if the `voltsmin` setting is _not_ null but is blank, which doesn't make sense to me.

tmleafs commented:
Wow should really go over my code more, I'm embarrassed with the amount of mistakes I have made on this pull request
I have changed the code to 

>     if(voltsmin == null || voltsmin == "")
>     	minVolts = 2.5
>     else
>    	minVolts = voltsmin
>     
>     if(voltsmax == null || voltsmax == "")
>     	maxVolts = 3.0
>     else 
>     maxVolts = voltsmax  
>
>

bspranger commented:
Ok, I have not looked over every line but what I did look at looks ok.

@veeceeoh have you looked over the changes?

I just want to be sure we don't break anything because I pissed off some ppl with a few of my commits.  🤣

veeceeoh commented:
> @veeceeoh have you looked over the changes?

I haven't looked at it in detail yet, no. I was only working on more compact code for determining the date format to use based on the user's preference setting. I came up with:

```
def formatDate(newBattery) {
def dateStyle = dateformat == "US” ? (newBattery ? “MMM dd yyyy” : "EEE MMM dd yyyy h:mm:ss a”) : (dateformat == "UK” ? (newBattery ? “dd MMM yyyy” : "EEE dd MMM yyyy h:mm:ss a”) : (newBattery ? “yyyy MMM dd” : "EEE yyyy MMM dd h:mm:ss a”))
return new Date().format(dateStyle, location.timeZone)
}
```
...but it's giving an error. 

Anyhow, I can test the code live with my sensors tonight, and will report back here.

veeceeoh commented:
Okay, I have done some testing, and the revised code for the Aqara Door/Window sensors is working correctly. 

However, the code for the ("original") Xiaomi Door/Window sensor must not be correct, because when I left the Date Format preference un-set (ie., I didn't choose one of the three styles), then the Battery Changed and Last Opened dates were displayed as YMD, while the Last Checkin date was displayed as MDY. Choosing one of the three date format options seems to fix those inconsistencies.

Also the override open function in the "original" Xiaomi Door/Window sensor DTH isn't updating the last opened date, and it should like the other DTHs.

I also tested the revised code for the Aqara Leak sensor, and that seems to be working all properly.

As for the battery min/max voltage preferences, I tried entering a bunch of illogical things. Here's what I did and what happened:

• Typed `3.`  --> upon save, it was changed to `3`
• Typed just a decimal point . --> No error, and battery reporting seems broken
• Typed low / high numbers --> Got errors: Min Volts must be at least 2 / no higher than 4
--> Got errors: Max Volts must be at least 2 / no higher than 5
• Typed `2` for both Min & Max volts --> no error, but battery reports coming in at 100%
• Typed `2` for Max and `4` for Min --> no error, battery reported at 50% twice, then stopped reporting

I don't know how it could be coded, but it would help to disallow entering in just a decimal point, and checking to make sure that Min/Max aren't the same number or Max is less than Min.

Unfortunately, I didn't have live logging open to see what was being reported there. I'll try doing that now. It takes a while because I have to wait for the normal 50/60 minute battery report intervals, and the infrequent check-in intervals.

veeceeoh commented:
Live Logging shows that:

• Entering in just a decimal point for either of the Min / Max volts preference settings results in a Groovy `MissingMethodException` error.
• Entering in a Max volts that is less than the Min volts setting seems to just result in an inversed percentage calculation. Example: raw battery of 3.065V is calculated to 47% based on Min volts of 4 and Max volts of 2. If those Min / Max values were swapped, the percentage would come out as 53%.
• If the Min / Max values are the same, I see a `Division by zero`error, and the raw voltage debug message isn't displayed at all. Not good.

tmleafs commented:
If we Change the range for min and max so they cant over lap or be the same that would solve most issues, I believe
So if min has a range of 2.0 to 2.7 and max is 2.8 to 3.4

alecm commented:
Commenting - 

a. Small typo "Only change the below settings if you know what your doing"  --> "Only change the below settings if you know what you're doing"   ("your" to "you're").

--------

b. Suggested change in wording - "Only change these settings if you know what you're doing."  or "Only change the settings below if you know what you're doing" 

c. Suggested change in wording
 
Max Volts - a battery is at 100% at ____ volts
Min Volts - a battery is at 0% (needs replacing) at ____ volts 

Last two are just suggestions

Anyhow I really like the idea of having this is an option.

tmleafs commented:
Yea, English isn't my strong point which is why I like computers and coding instead

tmleafs commented:
@alecm changes have been made 
@veeceeoh I have changed the ranges, doing so has stopped me from being able to put just a decimal point too 
![screenshot_20180124-152345 small](https://user-images.githubusercontent.com/8752233/35340788-080bf158-011c-11e8-9898-5957f154dcd8.jpg)

veeceeoh commented:
@tmleafs - nice! 

So it's pretty slow at work today, so I will look through the code in more detail to make sure it's consistent across all DTHs and check on why I was seeing inconsistencies in the date format presented for the ("original") Xiaomi Door/Window sensor DTH.

bspranger commented:
Should I hold off on merging this?

veeceeoh commented:
Well, based on my experience with testing last night I'm pretty confident the 3 DTHs I tested ("orig" Door/Window Sensor, Aqara Door/Window Sensor, & Aqara Leak Sensor) are working okay, except for the weird issue I mentioned with the ("original") Xiaomi Door/Window sensor DTH.

I haven't looked at the changes to any of the other DTHs, though.

Also, first thing I notice is at least on some of them, the preferences{} section is above metadata{} and I believe it should be the other way around.

veeceeoh commented:
So I went through everything and @tmleafs' code changes are consistent across all DTHs. The code formatting (indents) of the formatDate() function code needs to be fixed, but no big deal.

However, I found some related inconsistencies.

Most DTHs have these **4 lines** under the comment `// Send event for heartbeat`:
```
    //  send event for heartbeat
    def now = formatDate()    
    def nowDate = new Date(now).getTime()
    sendEvent(name: "lastCheckin", value: now)
    sendEvent(name: "lastCheckinDate", value: nowDate, displayed: false)
```

But the Xiaomi Aqara Motion Sensor and Xiaomi Motion Sensor DTHs only have two lines:
```
    def now = formatDate()    
    sendEvent(name: "lastCheckin", value: now)
```

...and the Xiaomi Aqara Temperature Humidity Sensor DTH only has 3 lines:
```
    def now = formatDate()   
    sendEvent(name: "lastCheckin", value: now)
    sendEvent(name: "lastCheckinDate", value: nowDate, displayed: false)
```

All the DTHs should be consistent in this send event for heartbeat section, right?

My last suggestion is to shorten the formatDate() code and change the variable “Boolean” to another more relevant name, like this:

```
def formatDate(batteryReset) {
    if (dateformat == "US" || dateformat == "" || dateformat == null) {
        if (batteryReset)
            return new Date().format("MMM dd yyyy", location.timeZone)
        else
            return new Date().format("EEE MMM dd yyyy h:mm:ss a", location.timeZone)
    }
    else if (dateformat == "UK") {
        if (batteryReset)
            return new Date().format("dd MMM yyyy", location.timeZone)
        else
            return new Date().format("EEE dd MMM yyyy h:mm:ss a", location.timeZone)
        }
    else {
        if (batteryReset)
            return new Date().format("yyyy MMM dd", location.timeZone)
        else
            return new Date().format("EEE yyyy MMM dd h:mm:ss a", location.timeZone)
    }
}
```

The only time the short date only format is used is when the user resets the battery time, so that’s why I used batteryReset. I also got rid of comparing batteryReset to “” and null by reversing the if / else order of the date format that gets returned.

I tested my modified formatDate() code with the DTHs for my sensors, and it works great.

bspranger commented:
I love the collaboration going on here!

Let's update the date function like described above and cleanup the heartbeat so it is consistent.   

Once those are done I will merge these in!

tmleafs commented:
All Done

veeceeoh commented:
Awesome! Thanks Steven, this is great.

bspranger commented:
Ok, we ready to go live?

veeceeoh commented:
Yup! Might be a good idea to announce it on your ST Community Forum thread.

veeceeoh commented:
Hey @tmleafs - how do you group different Commits on different files into one Pull Request like you did with this pull for the date format & max/min volts preference code?

tmleafs commented:
Whenever I add a feature make a new branch then if you make a pull request you can just keep adding to that branch and it will update the pull
